flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/table/RowDataKinesisDeserializationSchema.java

```diff
@@ -0,0 +1,122 @@
+ * physical data deserialization to an inner {@link DeserializationSchema} and appends requested
+ * metadata to the end of the deserialized {@link RowData} record.
+ */
+@Internal
+public class RowDataKinesisDeserializationSchema implements KinesisDeserializationSchema<RowData> {
+
+	/** Internal type for enumerating available metadata. */
+	protected enum Metadata {
+		ApproximateArrivalTimestamp(DataTypes.TIMESTAMP_WITH_LOCAL_TIME_ZONE(3).notNull()),
+		SequenceNumber(DataTypes.VARCHAR(128).notNull()),
+		ShardId(DataTypes.VARCHAR(128).notNull());
+
+		private final DataType dataType;
+
+		Metadata(DataType dataType) {
+			this.dataType = dataType;
+		}
+
+		public DataType getDataType() {
+			return this.dataType;
+		}
+	}
+
+	private static final long serialVersionUID = 5551095193778230749L;
+
+	/** A {@link DeserializationSchema} to deserialize the physical part of the row. */
+	private final DeserializationSchema<RowData> physicalDeserializer;
+
+	/** The type of the produced {@link RowData} records (physical data with appended metadata]. */
+	private final TypeInformation<RowData> producedTypeInfo;
+
+	/** Metadata fields to be appended to the physical {@link RowData} in the produced records. */
+	private final List<Metadata> requestedMetadataFields;
+
+	public RowDataKinesisDeserializationSchema(
+		DeserializationSchema<RowData> physicalDeserializer,
+		TypeInformation<RowData> producedTypeInfo,
+		List<Metadata> requestedMetadataFields) {
+		this.physicalDeserializer = Preconditions.checkNotNull(physicalDeserializer);
+		this.producedTypeInfo = Preconditions.checkNotNull(producedTypeInfo);
+		this.requestedMetadataFields = Preconditions.checkNotNull(requestedMetadataFields);
+	}
+
+	@Override
+	public void open(DeserializationSchema.InitializationContext context) throws Exception {
+		physicalDeserializer.open(context);
+	}
+
+	@Override
+	public RowData deserialize(
+		byte[] recordValue,
+		String partitionKey,
+		String seqNum,
+		long approxArrivalTimestamp,
+		String stream,
+		String shardId) throws IOException {
+
+		RowData physicalRow = physicalDeserializer.deserialize(recordValue);
+		GenericRowData metadataRow = new GenericRowData(requestedMetadataFields.size());
+
+		for (int i = 0; i < metadataRow.getArity(); i++) {
+			Metadata metadataField = requestedMetadataFields.get(i);
+			if (metadataField == Metadata.ApproximateArrivalTimestamp) {
+				metadataRow.setField(i, TimestampData.fromEpochMillis(approxArrivalTimestamp));
+			} else if (metadataField == Metadata.SequenceNumber) {
+				metadataRow.setField(i, StringData.fromString(seqNum));
+			} else if (metadataField == Metadata.ShardId) {
+				metadataRow.setField(i, StringData.fromString(shardId));
+			} else {
+				String msg = String.format("Unsupported metadata key %s", metadataField);
+				throw new RuntimeException(msg); // should never happen
+			}
+		}
+
+		JoinedRowData joinedRowData = new JoinedRowData(physicalRow, metadataRow);
```

Comment:
*Note*: I had to move `JoinedRowData` from `flink-table-runtime-blink` to `flink-table-common` to make this work. IMHO this is a more robust way of extending a record with metadata [than the current implementation in the `DynamicKafkaDeserializationSchema`](https://github.com/apache/flink/blob/d21f5d94474bd14eaa2d3c0dc386b81190533a4b/flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/table/DynamicKafkaDeserializationSchema.java#L115-L131).

Reply:
I factored this change in a separate commit.

I was hesitant to change the package name, though, as I am not sure whether this will be backwards-compatible. What will happen with savepoints that refer to serialized JoinedRowData instances after an 1.11.x → 1.12.0 upgrade? If you think this is safe I will change the package name.

Reply:
We do not provide any savepoint guarantees for SQL yet. We try to not change savepoints within minor versions but also this is just best effort. So we can safely move this class to a better package.

Reply:
OK, I added the relocation to a `utils` in a separate commit (different than the one that does the Maven module relocation). This will make it easier to revert the change and maintain backward-compatibility of the fully-qualified name without breaking the `RowDataKinesisDeserializationSchema` if needed.